### PR TITLE
introduced signals

### PR DESCRIPTION
**Pull request by asartori86**
I've just set a couple of functions that seemed reasonable to me. Any comments and suggestions for other signals?

### CODE REVIEW

**Comment by asartori86** [comment]
@luca-heltai @nicola-giuliani @ESeNonFossiIo ping

**Comment by ESeNonFossiIo** [comment]
Fine with me. I think we can merge this PR.

File: include/base_interface.h [hunk 1]
```diff
@@ -106,11 +110,24 @@ class BaseInterface : public ParameterAcceptor, public SimulatorAccess<dim,space
...
+   *         [this](VEC &y, VEC &y_dot)
+   *          {
+   *            y=1;
+   *            y_dot=0;
+   *          });
+   * @endcode
+   *
+   * An example of implementation is given in the poisson_problem_signals.h file.
+   */
+  virtual void connect_to_signals() const {}
```
Comment (luca-heltai) [inline include/base_interface.h]: Remove the `{}` from here, and put the default implementation in the cc file, or further down.

File: include/pidomus.h [hunk 2]
```diff
@@ -453,17 +454,15 @@ class piDoMUS : public ParameterAcceptor, public SundialsInterface<typename LAC:
...
-  /* /\** */
-  /*  * Distribute constraints at time t. This function update the */
-  /*  * functions and constraints at time t and then distribute them. */
-  /*  *\/ */
-  /* void distribute_constraints(const double &t) */
+
+  Signals<dim,spacedim,LAC>    signals;
 
   /**
    * SimulatoAccess accesses to all internal variables and returns a
```
Comment (luca-heltai) [inline include/pidomus.h]: SimulatorAccess

**Comment by luca-heltai** [comment]
Fix the above two comments, and merge. This looks fine to me. :+1:

**Comment by asartori86** [comment]
@luca-heltai  @ESeNonFossiIo Thanks for feedback! I was thinking to add a couple of signals with no arguments called at begin and end of each function of pidomus that can be used for debug (i.e. attaching a function that prints "I am here" or something like that). What do you think?

**Comment by luca-heltai** [comment]
I think this could be a good idea. Ideally, we should make sure each function in pi-domus is marked const, and modifies only its input-output parameters. In this way it would be easy to attach signals in a well defined manner, making the outcome reproducible...

**Comment by asartori86** [comment]
@luca-heltai @ESeNonFossiIo -- I've added the `begin_*` and `end_*` signals to pidomus. Look at the `poisson_problem_signals.h` for a possible use of these signals.

Let me paste a bit of the output when I run signals_01.cc

```
#########  make_grid_fe
#########  setup_dofs
Number of active cells: 4 (on 2 levels)
Number of degrees of freedom: 9(9)

ciao mondo
computing consistent initial conditions with the option use_y_diff please be patient.
#########  residual
#########  setup_jacobian
#########  assemble_matrices
#########  solve_jacobian_system
#########  residual
#########  setup_jacobian
#########  assemble_matrices
#########  solve_jacobian_system
compute initial conditions: done.
     0 ---->  0.01
#########  residual
#########  setup_jacobian
#########  assemble_matrices
#########  solve_jacobian_system
#########  residual
#########  solve_jacobian_system
#########  residual
#########  solve_jacobian_system
#########  residual
#########  solve_jacobian_system
#########  residual
#########  setup_jacobian
#########  assemble_matrices
#########  solve_jacobian_system
#########  residual
#########  setup_jacobian
#########  assemble_matrices
#########  solve_jacobian_system
#########  residual
#########  setup_jacobian
#########  assemble_matrices
#########  solve_jacobian_system
#########  residual
#########  setup_jacobian
#########  assemble_matrices
#########  solve_jacobian_system
#########  residual
#########  solve_jacobian_system
#########  residual
#########  setup_jacobian
#########  assemble_matrices
#########  solve_jacobian_system
#########  residual
#########  solve_jacobian_system
#########  solver_should_restart
  0.01 ---->  0.02
#########  residual
#########  solve_jacobian_system
#########  residual
#########  solve_jacobian_system
#########  solver_should_restart
  ################ restart ######### 
max_kelly > threshold
0.0194671 >  0.01
######################################
#########  refine_and_transfer_solutions
#########  setup_dofs
Number of active cells: 16 (on 3 levels)
Number of degrees of freedom: 25(25)

ho raffinato
computing consistent initial conditions with the option use_y_dot please be patient.
#########  residual
#########  setup_jacobian
#########  assemble_matrices
#########  solve_jacobian_system
#########  residual
#########  solve_jacobian_system
#########  residual
#########  setup_jacobian
#########  assemble_matrices
#########  solve_jacobian_system
compute initial conditions: done.
#########  solver_should_restart
```

**Pull request merged**